packages/medusa/src/services/discount-condition.ts

```diff
@@ -0,0 +1,216 @@
+      this.discountConditionRepository_
+    )
+
+    const query = this.buildQuery_({ id: conditionId }, config)
+    const condition = await conditionRepo.findOne(query)
+
+    if (!condition) {
+      throw new MedusaError(
+        MedusaError.Types.NOT_FOUND,
+        `DiscountCondition with id: ${conditionId} was not found`
+      )
+    }
+
+    return condition
+  }
+
+  resolveConditionType_(data: UpsertDiscountConditionInput):
+    | {
+        type: DiscountConditionType
+        resource_ids: string[]
+      }
+    | undefined {
+    switch (true) {
+      case !!data.products?.length:
+        return {
+          type: DiscountConditionType.PRODUCTS,
+          resource_ids: data.products!,
+        }
+      case !!data.product_collections?.length:
+        return {
+          type: DiscountConditionType.PRODUCT_COLLECTIONS,
+          resource_ids: data.product_collections!,
+        }
+      case !!data.product_types?.length:
+        return {
+          type: DiscountConditionType.PRODUCT_TYPES,
+          resource_ids: data.product_types!,
+        }
+      case !!data.product_tags?.length:
+        return {
+          type: DiscountConditionType.PRODUCT_TAGS,
+          resource_ids: data.product_tags!,
+        }
+      case !!data.customer_groups?.length:
+        return {
+          type: DiscountConditionType.CUSTOMER_GROUPS,
+          resource_ids: data.customer_groups!,
+        }
+      default:
+        return undefined
+    }
+  }
+
+  async upsertCondition(
+    discountId: string,
+    data: UpsertDiscountConditionInput
+  ): Promise<void> {
+    const resolvedConditionType = this.resolveConditionType_(data)
+
+    if (!resolvedConditionType) {
+      throw new MedusaError(
+        MedusaError.Types.INVALID_DATA,
+        `Missing one of products, collections, tags, types or customer groups in data`
+      )
+    }
+
+    return await this.atomicPhase_(
+      async (manager: EntityManager) => {
+        const discountConditionRepo: DiscountConditionRepository =
+          manager.getCustomRepository(this.discountConditionRepository_)
+        const discountRepo: DiscountRepository = manager.getCustomRepository(
+          this.discountRepository_
+        )
+
+        if (data.id) {
+          // throw if condition does not exist
+          const condition = await this.retrieve(data.id)
+
+          return await discountConditionRepo.addConditionResources(
+            condition.id,
+            resolvedConditionType.resource_ids,
+            resolvedConditionType.type,
+            true
+          )
+        }
+
+        const discount = await discountRepo.findOne({
+          where: { id: discountId },
+          relations: ["rule", "rule.conditions"],
+        })
+
+        if (!discount) {
+          throw new MedusaError(
+            MedusaError.Types.NOT_FOUND,
+            `Discount with ${discountId} was not found`
+          )
+        }
+
+        const created = discountConditionRepo.create({
```

Comment:
`createdDiscountCondition` We should formalise the wording. The best practice says that the code must be readable as a book, the wording is important everywhere :)